### ISSUE TITLE
Uncaught exception using ivy

### ISSUE THREAD

**Issue by fr0** [angular/angular#31314]
# 🐞 bug report

### Affected Package

Unknown.

### Is this a regression?

Only happen when using Ivy.

### Description

An unhandled exception occurs immediately, and my app does not load.

## 🔥 Exception or Error
<pre><code>
core.js:30223 Uncaught TypeError: Cannot read property 'id' of undefined
    at registerNgModuleType (core.js:30223)
    at core.js:30241
    at Array.forEach (<anonymous>)
    at registerNgModuleType (core.js:30237)
    at new NgModuleFactory$1 (core.js:30375)
    at compileNgModuleFactory__POST_R3__ (core.js:34663)
    at PlatformRef.bootstrapModule (core.js:34960)
    at Module../src/main.ts (main.ts:11)
    at __webpack_require__ (bootstrap:79)
    at Object.0 (main.ts:11)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 8.1.0-beta.3
Node: 12.4.0
OS: darwin x64
Angular: 8.1.0-rc.0
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.801.0-beta.3
@angular-devkit/build-angular     0.801.0-beta.3
@angular-devkit/build-optimizer   0.801.0-beta.3
@angular-devkit/build-webpack     0.801.0-beta.3
@angular-devkit/core              8.1.0-beta.3
@angular-devkit/schematics        8.1.0-beta.3
@angular/cli                      8.1.0-beta.3
@ngtools/json-schema              1.1.0
@ngtools/webpack                  8.1.0-beta.3
@schematics/angular               8.1.0-beta.3
@schematics/update                0.801.0-beta.3
ng-packagr                        5.3.0
rxjs                              6.5.2
typescript                        3.4.5
webpack                           4.35.0
</code></pre>

**Comment by AndrewKushnir** [comment]
@fr0, could you please provide some additional information on how to reproduce this issue? It'd be helpful if you could create a repro using https://stackblitz.com/, so we can investigate the problem. Thank you.

**Comment by bennadel** [comment]
I get the same error when I try to import the Animations module. I don't even attempt to use it -- this is _just_ from the import:

```ts
import { BrowserAnimationsModule } from "@angular/platform-browser/animations";
import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";

@NgModule({
	imports: [
		BrowserModule,
		BrowserAnimationsModule
	],
	....
```

![2019-06-28_06-41-32](https://user-images.githubusercontent.com/563690/60336988-d118e500-996f-11e9-81e2-b7a123254468.png)

**Comment by fr0** [comment]
Removing `BrowserAnimationsModule` does "fix" the problem for me as well.

**Comment by fr0** [comment]
While trying to create a reproducible repo, I added back the `BrowserAnimationsModule` and now I'm getting a different (compile-time) error:

```
ERROR in ../node_modules/@angular/platform-browser/animations/animations.d.ts:37:22 - error TS-996002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

37 export declare class BrowserAnimationsModule {
                        ~~~~~~~~~~~~~~~~~~~~~~~
```
I have no idea what I did wrong. I'll poke more at this Monday.

**Comment by AndrewKushnir** [comment]
Thanks for additional information.

It looks like this issue is similar to the one described in #31348 and there is a fix that should resolve the issue: https://github.com/angular/angular-cli/pull/14934, [see comment](https://github.com/angular/angular/issues/31348#issuecomment-506920483) from @alexzuza. The fix will most likely be included into the next CLI release.

**Comment by alxhub** [comment]
Fixed by the commit mentioned above.

**Comment by lwensveen** [comment]
Still get this issue with 8.2.0-next.0

Example:
`node_modules/@angular/common/common.d.ts(115,22): error TS-996002: Appears in the NgModule.imports of CoreModule, but could not be resolved to an NgModule class`

**Comment by hoang-innomizetech** [comment]
I faced this exception, then I have to turn off Ivy option to get detailed error

```
 "angularCompilerOptions": {
    "enableIvy": true
  }
```

**Comment by alexzuza** [comment]
@lwensveen Can you please provide a minimal reproducible example of your case?

**Comment by bennadel** [comment]
I still get this in error in `8.2.0-next.0` as well. And, what I realize is that it is not specific to the **Animations** module. I was also getting it when I tried to pull in the **HTTPClient** module as well. I wonder if it is any module other than the base browser one?

**Comment by spock123** [comment]
Getting same here with 8.2.0-next.2

> Cannot read property 'id' of undefined

**Comment by sviat9440** [comment]
Getting the same with 8.2.2

**Comment by eabassey** [comment]
Getting the same error with 8.1.0

**Comment by Anyoks** [comment]
Getting the same issue with

```
Angular CLI: 8.0.6
Node: 11.10.0
OS: linux x64
Angular: 9.0.0-next.6

@angular-devkit/architect         0.803.4
@angular-devkit/build-angular     0.803.4
@angular-devkit/build-optimizer   0.803.4
@angular-devkit/build-webpack     0.803.4
@angular-devkit/core              8.3.4
@angular-devkit/schematics        8.0.6
@angular/cdk                      8.2.0
@angular/cli                      8.0.6
@angular/http                     6.1.10
@angular/material                 8.2.0
@ngtools/webpack                  8.3.4
@schematics/angular               8.0.6
@schematics/update                0.800.6
rxjs                              6.5.3
typescript                        3.5.3
webpack                           4.39.2

```

**Comment by haerulmutakin** [comment]
getting same error with 8.2.7. any solution?

**Comment by hugoebarboza** [comment]
getting same error any solution?

**Comment by gkalpak** [comment]
Can someone, please, provide an (ideally minimal) reproduction. Otherwise, we can't investigate and get to the bottom of the issue.

Also, note that most Ivy improvements/fixes are not backported to 8.x, so make worth checking whether the issue still happens with the latest 9.x release (because the bug might already have been fixed there).

**Comment by WilliamSf** [comment]
`Angular CLI: 8.3.6
Node: 10.16.3
OS: linux x64
Angular: 9.0.0-next.8
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.6
@angular-devkit/build-angular     0.803.6
@angular-devkit/build-optimizer   0.803.6
@angular-devkit/build-webpack     0.803.6
@angular-devkit/core              8.3.6
@angular-devkit/schematics        8.3.6
@angular/cdk                      8.2.2
@angular/cli                      8.3.6
@angular/flex-layout              8.0.0-beta.27
@angular/material                 8.2.2
@ngtools/webpack                  8.3.6
@schematics/angular               8.3.6
@schematics/update                0.803.6
rxjs                              6.5.3
typescript                        3.5.3
webpack                           4.39.2
`


package.json 
`"dependencies": {
    "@angular/animations": "~9.0.0-next.8",
    "@angular/cdk": "^8.2.2",
    "@angular/common": "~9.0.0-next.8",
    "@angular/compiler": "~9.0.0-next.8",
    "@angular/core": "~9.0.0-next.8",
    "@angular/flex-layout": "^8.0.0-beta.27",
    "@angular/forms": "~9.0.0-next.8",
    "@angular/material": "^8.2.2",
    "@angular/platform-browser": "~9.0.0-next.8",
    "@angular/platform-browser-dynamic": "~9.0.0-next.8",
    "@angular/router": "~9.0.0-next.8",
    "hammerjs": "^2.0.8",
    "json-server": "^0.15.1",
    "module-alias": "^2.2.1",
    "moment": "^2.24.0",
    "ngx-file-drop": "^8.0.7",
    "rxjs": "~6.5.3",
    "stacktrace-js": "^2.0.1",
    "tslib": "^1.9.0",
    "zone.js": "~0.10.2"
  }`

error:
`Uncaught TypeError: Cannot read property 'id' of undefined
    at registerNgModuleType (core.js:34738)
    at core.js:34756
    at Array.forEach (<anonymous>)
    at registerNgModuleType (core.js:34752)
    at core.js:34756
    at Array.forEach (<anonymous>)
    at registerNgModuleType (core.js:34752)
    at new NgModuleFactory$1 (core.js:34919)
    at compileNgModuleFactory__POST_R3__ (core.js:40668)
    at PlatformRef.bootstrapModule (core.js:41010)`

Any solution ?

**Comment by hugoebarboza** [comment]
same error for my, any solution?

**Comment by WilliamSf** [comment]
no :/

**Comment by shankarajabe** [comment]
Same error

core.js:34469 Uncaught TypeError: Cannot read property 'id' of undefined
    at registerNgModuleType (core.js:34469)
    at core.js:34487
    at Array.forEach (<anonymous>)
    at registerNgModuleType (core.js:34483)
    at new NgModuleFactory$1 (core.js:34650)
    at compileNgModuleFactory__POST_R3__ (core.js:40285)
    at PlatformRef.bootstrapModule (core.js:40608)
    at Module../src/main.ts (main.ts:11)
    at __webpack_require__ (bootstrap:84)
    at Object.0 (main.ts:12)

**Comment by Splaktar** [comment]
@gkalpak @alexzuza I hit this as well in a non-CLI project. Here's a branch with the repo: https://github.com/DevIntent/material2/tree/ivy-repro-31314.

My guess atm is that we're using `bootstrapModuleFactory` which doesn't work in Ivy? I'm still looking into it...

**Comment by Splaktar** [comment]
For the reproduction that I posted above, I believe that our solution to this was changing [from](https://github.com/DevIntent/material2/blob/ivy-repro-31314/package.json#L16)
```sh
ivy-ngcc --properties es2015 --create-ivy-entry-points
```
[to](https://github.com/angular/components/blob/8f0f73c4237a9c5793df55014d2cfac964f27c3a/package.json#L16)
```sh
ivy-ngcc --properties main
```
Please note that `ivy-ngcc` has since been renamed to `ngcc`.

**Comment by alxhub** [comment]
Thanks @Splaktar! Closing as resolved.

**Comment by fclemonschool** [comment]
angular version 9.0.0-rc.6 error still occurs.

**Comment by Splaktar** [comment]
@fclemonschool please provide a reproduction via a GitHub repo. This can be caused be a misconfiguration of a project, so we need to see the full project to debug.

**Comment by DenisKudelin** [comment]
The same issue, I think it's related to lazy module loading using routes

**Comment by vfe12345** [comment]
Same problem. @angular/cdk did not get compiled properly.
Whole folder __ivy_ngcc__ is missing in __ivy_ngcc__. That's why this exception occurs during runtime.
Compiler was happy all the time.... no exception there... only during runtime.

Then I tried to add postinstall script ... ngcc to pre-build this. It failed in one of @angular/xyz assemblies.

I am using yarn + lerna. This causes trouble every single time I try to update.

I created a repository for specific people to investigate (minimum version extracted from existing project)...
https://github.com/vfe12345/IvyProblem.git
